File: trunk/simulations-java/simulations/beers-law-lab/src/edu/colorado/phet/beerslawlab/view/MovableDragHandler.java
// Copyright 2002-2011, University of Colorado
package edu.colorado.phet.beerslawlab.view;

import java.awt.geom.Point2D;
import java.awt.geom.Rectangle2D;

import edu.colorado.phet.beerslawlab.BLLSimSharing.Parameters;
import edu.colorado.phet.beerslawlab.model.Movable;
import edu.colorado.phet.common.phetcommon.math.ImmutableVector2D;
import edu.colorado.phet.common.phetcommon.simsharing.Parameter;
import edu.colorado.phet.common.phetcommon.simsharing.SimSharingManager;
import edu.colorado.phet.common.phetcommon.simsharing.messages.ParameterSet;
import edu.colorado.phet.common.phetcommon.simsharing.messages.UserAction;
import edu.colorado.phet.common.phetcommon.simsharing.messages.UserComponent;
import edu.colorado.phet.common.phetcommon.util.function.Function0;
import edu.colorado.phet.common.piccolophet.simsharing.SimSharingDragSequenceEventHandler;
import edu.umd.cs.piccolo.PNode;
import edu.umd.cs.piccolo.event.PInputEvent;

/**
 * A drag handler for something that is movable.
 *
 * @author Chris Malley (cmalley@pixelzoom.com)
 */
public class MovableDragHandler extends SimSharingDragSequenceEventHandler {

    private final Movable movable;
    private final PNode dragNode;
    private double clickXOffset, clickYOffset; // offset of mouse click from dragNode's origin, in parent's coordinate frame

    public MovableDragHandler( final UserComponent userComponent, final Movable movable, PNode dragNode ) {
        this( userComponent, movable, dragNode,
              new Function0<Parameter[]>() {
                  public Parameter[] apply() {
                      return new Parameter[0];
                  }
              } );
    }

    public MovableDragHandler( final UserComponent userComponent, final Movable movable, PNode dragNode, final Function0<Parameter[]> simSharingParametersFunction ) {
        this.movable = movable;
        this.dragNode = dragNode;
        // sim-sharing
        setStartEndDragFunction( new DragFunction() {
            public void apply( UserAction action, Parameter xParameter, Parameter yParameter, PInputEvent event ) {
                SimSharingManager.sendUserMessage( userComponent, action,
                                                   new ParameterSet( simSharingParametersFunction.apply() ).param( Parameters.locationX, (int) movable.location.get().getX() ).
                                                           param( Parameters.locationY, (int) movable.location.get().getY() ) );
            }
        } );
    }

    @Override protected void startDrag( PInputEvent event ) {
        super.startDrag( event );
        Point2D pMouse = event.getPositionRelativeTo( dragNode.getParent() );
        clickXOffset = pMouse.getX() - movable.location.get().getX();
        clickYOffset = pMouse.getY() - movable.location.get().getY();
    }

    @Override protected void drag( final PInputEvent event ) {
        super.drag( event );
        Point2D pMouse = event.getPositionRelativeTo( dragNode.getParent() );
        double x = pMouse.getX() - clickXOffset;
        double y = pMouse.getY() - clickYOffset;
        //TODO assumes a 1:1 model-view transform
        movable.location.set( constrainToBounds( x, y, movable.getDragBounds() ) );
    }

    // Constrains xy coordinates to be within the specified bounds.
    private static ImmutableVector2D constrainToBounds( double x, double y, Rectangle2D bounds ) {
        ImmutableVector2D vConstrained;
        if ( bounds != null && !bounds.contains( x, y ) ) {
            double xConstrained = Math.max( Math.min( x, bounds.getMaxX() ), bounds.getX() );
            double yConstrained = Math.max( Math.min( y, bounds.getMaxY() ), bounds.getY() );
            vConstrained = new ImmutableVector2D( xConstrained, yConstrained );
        }
        else {
            vConstrained = new ImmutableVector2D( x, y );
        }
        return vConstrained;
    }
}

File: simulations-html/simulations/beers-law-lab/js/common/view/MovableDragHandler.js
// Copyright 2002-2013, University of Colorado

/**
 * A drag handler for something that is movable and constrained to some bounds.
 * All values herein are in the view coordinate frame.
 *
 * @author Chris Malley (cmalley@pixelzoom.com)
 */
define( [
            'phetcommon/math/Point2D'
        ],
        function ( Point2D ) {

            function MovableDragHandler() {
            }

            /**
             * Registers a drag handler with the specified Easel display object.
             * @param {DisplayObject} dragNode
             * @param {Rectangle} dragBounds
             * @param {Function} dragFunction
             */
            MovableDragHandler.register = function ( dragNode, dragBounds, dragFunction ) {

                // Drag cursor
                dragNode.onMouseOver = function () {
                    document.body.style.cursor = "pointer";
                };

                // Normal cursor
                dragNode.onMouseOut = function () {
                    document.body.style.cursor = "default";
                };

                // @param {MouseEvent} pressEvent
                dragNode.onPress = function ( pressEvent ) {

                    // Make dragging relative to touch Point2D.
                    var relativePressPoint = null;

                    // @param {MouseEvent} moveEvent
                    pressEvent.onMouseMove = function ( moveEvent ) {
                        var transformed = moveEvent.target.parent.globalToLocal( moveEvent.stageX, moveEvent.stageY );
                        if ( relativePressPoint === null ) {
                            relativePressPoint = new Point2D( pressEvent.target.x - transformed.x, pressEvent.target.y - transformed.y );
                        }
                        else {
                            var p = new Point2D( transformed.x + relativePressPoint.x, transformed.y + relativePressPoint.y );
                            dragFunction( MovableDragHandler.constrainBounds( p, dragBounds ) );
                        }
                    };
                };
            };

            /**
             * Constrains a point to some bounds.
             * @param {Point2D} point
             * @param {Rectangle} bounds
             */
            MovableDragHandler.constrainBounds = function ( point, bounds ) {
                if ( bounds === undefined || bounds.contains( point.x, point.y ) ) {
                    return point;
                }
                else {
                    var xConstrained = Math.max( Math.min( x, bounds.getMaxX() ), bounds.x );
                    var yConstrained = Math.max( Math.min( y, bounds.getMaxY() ), bounds.y );
                    return new Point2D( xConstrained, yConstrained );
                }
            };

            return MovableDragHandler;
        } );
